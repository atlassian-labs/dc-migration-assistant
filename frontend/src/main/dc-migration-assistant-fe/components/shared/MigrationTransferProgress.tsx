import React, { ReactElement, FunctionComponent } from 'react';
import { Moment } from 'moment';
import SectionMessage from '@atlaskit/section-message';
import Spinner from '@atlaskit/spinner';
import ProgressBar, { SuccessProgressBar } from '@atlaskit/progress-bar';
import { I18n } from '../../atlassian/mocks/@atlassian/wrm-react-i18n';

import { Progress } from './Progress';
import {
    calculateDurationFromBeginning,
    calcualateDurationFromElapsedSeconds,
    calculateStartedFromElapsedSeconds,
} from './migration-timing';

export type MigrationProgressProps = {
    progress: Progress;
    /**
     * Should be true when the progress is being fetched
     */
    loading: boolean;
    /**
     * A moment representing the time that the transfer was started
     */
    startedMoment: Moment;
};

/**
 * A component which renders the progress of the current transfer. It renders
 * a progress bar and elapsed time. It handles when the completeness is indeterminate.
 */
export const MigrationProgress: FunctionComponent<MigrationProgressProps> = ({
    progress,
    loading,
    startedMoment,
}) => {
    if (loading) {
        return (
            <>
                <Spinner />
                <ProgressBar isIndeterminate />
                <Spinner />
            </>
        );
    }

    const duration =
        calculateDurationFromBeginning(startedMoment) ||
        calcualateDurationFromElapsedSeconds(progress.elapsedTimeSeconds);

    return (
        <>
            {progress?.completeness === 1 && progress?.completeMessage && (
                <SectionMessage appearance="confirmation">
                    <strong>{progress.completeMessage.boldPrefix}</strong>{' '}
                    {progress.completeMessage.message}
                </SectionMessage>
            )}

            <>
                <h4>
                    {progress.phase}
                    {progress.completeness === undefined &&
                        ` (${I18n.getText('atlassian.migration.datacenter.common.estimating')}...)`}
                </h4>
                {progress.completeness ? (
                    <SuccessProgressBar value={progress.completeness} />
                ) : (
                    <ProgressBar isIndeterminate />
                )}
                <p>
                    {I18n.getText(
                        'atlassian.migration.datacenter.common.progress.started',
                        (
                            startedMoment ||
                            calculateStartedFromElapsedSeconds(progress.elapsedTimeSeconds)
                        ).format('D/MMM/YY h:mm A')
                    )}
                </p>
                <p>
                    {duration &&
                        I18n.getText(
                            'atlassian.migration.datacenter.common.progress.mins_elapsed',
                            `${duration.days * 24 + duration.hours}`,
                            `${duration.minutes}`
                        )}
                </p>
            </>
        </>
    );
};
