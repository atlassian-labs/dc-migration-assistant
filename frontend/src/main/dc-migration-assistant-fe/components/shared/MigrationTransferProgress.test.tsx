import React from 'react';
import { render } from '@testing-library/react';
import { MigrationProgress } from './MigrationTransferProgress';

const defaultCompleteMessage = {
    boldPrefix: '180GB of database',
    message: 'were transferred successfully',
};

describe('migration process progress', () => {
    it('Should render indeterminate progress bar when completeness is not given', () => {
        const { container } = render(
            <MigrationProgress
                loading={false}
                progress={{
                    elapsedTimeSeconds: 120,
                    phase: 'Transferring your data. This may take a while...',
                    completeMessage: defaultCompleteMessage,
                }}
            />
        );

        expect(container).toMatchSnapshot();
    });

    it('Should render a partially completed progress bar with timing data when completeness is between 0 and 1', () => {
        const { container } = render(
            <MigrationProgress
                loading={false}
                progress={{
                    phase: 'Downloading data into new instance',
                    completeness: 0.5,
                    elapsedTimeSeconds: 360,
                    completeMessage: defaultCompleteMessage,
                }}
            />
        );

        expect(container).toMatchSnapshot();
    });

    it('Should render an error message when one is given', () => {
        const { container } = render(
            <MigrationProgress
                loading={false}
                progress={{
                    phase: 'Failed to install data in new instance',
                    completeness: 0.7,
                    elapsedTimeSeconds: 720,
                    errorMessage: 'Unable to connect to target database',
                    completeMessage: defaultCompleteMessage,
                }}
            />
        );

        expect(container).toMatchSnapshot();
    });

    it('Should render a full progress bar when process completeness is 1', () => {
        const { container } = render(
            <MigrationProgress
                loading={false}
                progress={{
                    phase: 'Complete',
                    completeness: 1,
                    elapsedTimeSeconds: 1080,
                    completeMessage: defaultCompleteMessage,
                }}
            />
        );

        expect(container).toMatchSnapshot();
    });
});
