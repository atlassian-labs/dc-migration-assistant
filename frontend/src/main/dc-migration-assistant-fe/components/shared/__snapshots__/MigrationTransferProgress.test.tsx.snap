// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migration process progress Should render a full progress bar when process completeness is 1 1`] = `
<div>
  <section
    class="sc-bxivhb brjzuU"
  >
    <div
      class="sc-dnqmqq PgEDm"
    >
      <span
        class="sc-htpNat jsDFjY"
        role="presentation"
      >
        <svg
          focusable="false"
          height="24"
          role="presentation"
          viewBox="0 0 24 24"
          width="24"
        >
          <g
            fill-rule="evenodd"
          >
            <circle
              cx="12"
              cy="12"
              fill="currentColor"
              r="10"
            />
            <path
              d="M9.707 11.293a1 1 0 1 0-1.414 1.414l2 2a1 1 0 0 0 1.414 0l4-4a1 1 0 1 0-1.414-1.414L11 12.586l-1.293-1.293z"
              fill="inherit"
            />
          </g>
        </svg>
      </span>
    </div>
    <div
      class="sc-ifAKCX ezQlwD"
    >
      <div
        class="sc-bZQynM jBPlWM"
      >
        <strong>
          180GB of database
        </strong>
         
        were transferred successfully
      </div>
    </div>
  </section>
  <h4
    class="sc-gZMcBi jPsQyQ"
  >
    Complete
  </h4>
  <div
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow="1"
    class="css-1656qb0"
    role="progressbar"
    tabindex="0"
  >
    <span
      class="css-9m5vj-Bar"
    />
  </div>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    atlassian.migration.datacenter.common.progress.started 21/Jul/20 4:03 PM
  </p>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    0 atlassian.migration.datacenter.common.progress.hours, 18 atlassian.migration.datacenter.common.progress.minutes, 0 atlassian.migration.datacenter.common.progress.seconds atlassian.migration.datacenter.common.progress.elapsed
  </p>
</div>
`;

exports[`migration process progress Should render a partially completed progress bar with timing data when completeness is between 0 and 1 1`] = `
<div>
  <h4
    class="sc-gZMcBi jPsQyQ"
  >
    Downloading data into new instance
  </h4>
  <div
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow="0.5"
    class="css-1656qb0"
    role="progressbar"
    tabindex="0"
  >
    <span
      class="css-9b3agr-Bar"
    />
  </div>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    atlassian.migration.datacenter.common.progress.started 21/Jul/20 4:15 PM
  </p>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    0 atlassian.migration.datacenter.common.progress.hours, 6 atlassian.migration.datacenter.common.progress.minutes, 0 atlassian.migration.datacenter.common.progress.seconds atlassian.migration.datacenter.common.progress.elapsed
  </p>
</div>
`;

exports[`migration process progress Should render an error message when one is given 1`] = `
<div>
  <section
    class="sc-bxivhb jsxfHf"
  >
    <div
      class="sc-dnqmqq PgEDm"
    >
      <span
        class="sc-htpNat kNbdZf"
        role="presentation"
      >
        <svg
          focusable="false"
          height="24"
          role="presentation"
          viewBox="0 0 24 24"
          width="24"
        >
          <g
            fill-rule="evenodd"
          >
            <path
              d="M13.416 4.417a2.002 2.002 0 0 0-2.832 0l-6.168 6.167a2.002 2.002 0 0 0 0 2.833l6.168 6.167a2.002 2.002 0 0 0 2.832 0l6.168-6.167a2.002 2.002 0 0 0 0-2.833l-6.168-6.167z"
              fill="currentColor"
            />
            <path
              d="M12 14a1 1 0 0 1-1-1V8a1 1 0 0 1 2 0v5a1 1 0 0 1-1 1m0 3a1 1 0 0 1 0-2 1 1 0 0 1 0 2"
              fill="inherit"
            />
          </g>
        </svg>
      </span>
    </div>
    <div
      class="sc-ifAKCX ezQlwD"
    >
      <div
        class="sc-bZQynM jBPlWM"
      >
        <p
          class="sc-iwsKbI fQqfNj"
        >
          Unable to connect to target database
           
          <a
            href="https://confluence.atlassian.com/jirakb/how-to-use-the-data-center-migration-app-to-migrate-jira-to-an-aws-cluster-1005781495.html#HowtousetheDataCenterMigrationapptomigrateJiratoanAWScluster-errors"
            rel="noreferrer noopener"
            target="_blank"
          >
            atlassian.migration.datacenter.common.learn_more
          </a>
        </p>
      </div>
    </div>
  </section>
  <h4
    class="sc-gZMcBi jPsQyQ"
  >
    Failed to install data in new instance
  </h4>
  <div
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow="0.7"
    class="css-1656qb0"
    role="progressbar"
    tabindex="0"
  >
    <span
      class="css-bpkja2-Bar"
    />
  </div>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    atlassian.migration.datacenter.common.progress.started 21/Jul/20 4:09 PM
  </p>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    0 atlassian.migration.datacenter.common.progress.hours, 12 atlassian.migration.datacenter.common.progress.minutes, 0 atlassian.migration.datacenter.common.progress.seconds atlassian.migration.datacenter.common.progress.elapsed
  </p>
</div>
`;

exports[`migration process progress Should render indeterminate progress bar when completeness is not given 1`] = `
<div>
  <h4
    class="sc-gZMcBi jPsQyQ"
  >
    Transferring your data. This may take a while...
     (atlassian.migration.datacenter.common.estimating...)
  </h4>
  <div
    aria-valuemax="1"
    aria-valuemin="0"
    aria-valuenow="0"
    class="css-1656qb0"
    role="progressbar"
    tabindex="0"
  >
    <span
      class="css-1w0nw0i-Bar"
    />
    <span
      class="css-1xfl40c-Bar"
    />
  </div>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    atlassian.migration.datacenter.common.progress.started 21/Jul/20 4:19 PM
  </p>
  <p
    class="sc-gqjmRU kwvhvQ"
  >
    0 atlassian.migration.datacenter.common.progress.hours, 2 atlassian.migration.datacenter.common.progress.minutes, 0 atlassian.migration.datacenter.common.progress.seconds atlassian.migration.datacenter.common.progress.elapsed
  </p>
</div>
`;
